pallets/pallet-lottery/src/lib.rs

```diff
@@ -0,0 +1,309 @@
+		pub fn update_ticket_price(origin: OriginFor<T>, new_price: T::Balance) -> DispatchResult {
+			let id = ensure_signed(origin)?;
+			let role = T::RoleManager::role(&id);
+			ensure!(
+				role == Some(Role::Manager) || role == Some(Role::Auditor),
+				Error::<T>::IncorrectRole
+			);
+			let old = TicketPrice::<T>::get();
+			// Update the ticket price
+			TicketPrice::<T>::put(new_price);
+			Self::deposit_event(Event::<T>::TicketPriceUpdated { old, new: new_price });
+
+			Ok(())
+		}
+
+		/// Buy numbers of tickets and transfer the funds from the customer account to the lottery
+		/// pool account.
+		///
+		/// Required Customer.
+		#[pallet::call_index(2)]
+		#[pallet::weight(T::WeightInfo::buy_ticket())]
+		pub fn buy_ticket(origin: OriginFor<T>, number_of_tickets: u32) -> DispatchResult {
+			// Ensure the caller is a customer account
+			let id = ensure_signed(origin)?;
+			T::RoleManager::ensure_role(&id, Role::Customer)?;
+
+			// Calculate the total price of the tickets
+			ensure!(!TicketPrice::<T>::get().is_zero(), Error::<T>::TicketPriceNotSet);
+			let total_price = TicketPrice::<T>::get().saturating_mul(number_of_tickets.into());
+
+			// Transfer total_price from the customer to the PrizePoolAccount.
+			T::Bank::transfer(&id, &T::PrizePoolAccount::get(), total_price)?;
+
+			// Check if the player exists
+			if let Some(mut tickets) = PlayersAndLotteries::<T>::get(id.clone()) {
+				// Player exists, update the number of tickets
+				tickets = tickets.saturating_add(number_of_tickets);
+				// Update the value in the storage map
+				PlayersAndLotteries::<T>::insert(id.clone(), tickets);
+			} else {
+				// Player doesn't exist, insert a new entry with the given number of tickets
+				PlayersAndLotteries::<T>::insert(id.clone(), number_of_tickets);
+			}
+			Self::deposit_event(Event::<T>::TicketsBought {
+				id: id.clone(),
+				number: number_of_tickets,
+			});
+
+			Ok(())
+		}
+	}
+
+	impl<T: Config> Pallet<T> {
+		/// Checks if the prize split adds up to 100% exactly.
+		fn check_split_valid(split: &[Percent]) -> bool {
+			split.iter().map(|x| *x * 100u32).sum::<u32>() == 100u32
+		}
+
+		/// Increase the random seed.
+		fn next_seed() -> Vec<u8> {
+			let seed = RandomSeed::<T>::get();
+			RandomSeed::<T>::put(seed.wrapping_add(1));
+			seed.encode()
+		}
+
+		/// Payout the won fund to the winner and paid tax to the treasury account.
+		fn resolve_lottery_winner() {
+			if let Ok(treasury) = T::Bank::treasury() {
+				// Set up data for choosing winners
+				let total = T::Bank::free_balance(&T::PrizePoolAccount::get());
+				let tax_rate = T::TaxRate::get();
+				let number_of_winners = PrizeSplit::<T>::get().len();
+				let players = PlayersAndLotteries::<T>::iter().collect::<Vec<_>>();
+
+				// Choose the winners
+				let winners = Self::select_n_winners(players, number_of_winners as u32);
+
+				for (i, user) in winners.into_iter().enumerate() {
+					let percent = PrizeSplit::<T>::get()[i];
```

Comment:
storage gets are expensive. We should get it once and store it locally. 

`let prizes = PrizeSplit::<T>::get().map(|p| p * total).collect::<Vec<_>>();`